### PR TITLE
Support HTTPS via user-provided key/cert

### LINKED ISSUE #244: Support SSL key/cert files
#229 adds handy support for SSL via letsencrypt. We should have similarly simple support when users have existing SSL key/cert files. config should probably look something like:

```yaml
proxy:
  https:
    hosts:
      - needed?
    kind: files
    files:
      key: |
        -----BEGIN RSA PRIVATE KEY-----
        ...
        -----END RSA PRIVATE KEY-----
      cert: |
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
```

### PR DESCRIPTION
**Pull request by minrk**
usage:

```yaml
proxy:
  https:
    type: files
    files:
      cert: |
        <paste cert>
      key: |
        <paste key>
```

closes #244

### CODE REVIEW

File: jupyterhub/Chart.yaml [hunk 1]
```diff
@@ -1,4 +1,4 @@
 apiVersion: v1
 description: Multi-user Jupyter installation
 name: jupyterhub
-version: v0.5.0
+version: v0.5.0-3be4d58
```
Comment (yuvipanda) [inline jupyterhub/Chart.yaml]: Accidental?

**Comment by yuvipanda** [comment]
This is awesome!

What do you think about calling this 'manual' or something like that rather than 'files'? Am unsure what 'files' signifies in this context...

**Comment by minrk** [comment]
I was thinking 'files' because they came from files, but of course they aren't files anymore when they go in your .yml. manual seems okay, but it's not super obvious what it means.

Does it make sense for the default behavior to use `https.key/cert` if defined, or is that too many choices to select between automatically?

File: doc/source/extending-jupyterhub.rst [hunk 2]
```diff
@@ -198,26 +198,58 @@ your users & the world at large. Zero to JupyterHub makes doing so quite
 easy since version 0.5, integrating with `Let's Encrypt <https://letsencrypt.org/>`_
 for free HTTPS certificates.
 
+You can also purchase your own SSL certificates from a certificate provider.
+
 1. Buy a domain name from a registrar. Pick whichever one you want.
 2. Create an ``A record`` from the domain you want to use, pointing to the
    external IP provided to the `proxy-public` service.
```
Comment (choldgraf) [inline doc/source/extending-jupyterhub.rst]: could explicitly say this is the `<EXTERNAL-IP>` field. For whatever reason people almost always get this confused and copypaste the internal IP

Comment (choldgraf) [reply doc/source/extending-jupyterhub.rst]: (I know this isn't related to this PR so we can punt it to later if you like)

File: jupyterhub/templates/proxy/ingress-internal.yaml [hunk 3]
```diff
@@ -24,8 +23,15 @@ spec:
...
     {{- end }}
   tls:
+    {{- if eq .Values.proxy.https.type "letsencrypt" }}
     - secretName: kubelego-tls-proxy-{{ .Release.Name }}
+    {{- else if eq .Values.proxy.https.type "files" }}
+    - secretName: files-tls-proxy-{{ .Release.Name }}
+    {{- else }}
+    # unhandled type
+    # it would be nice if helm had an `error` function
+    {{ required (printf "https.type must be 'files' or 'letsencrypt', not '%s'" .Values.proxy.https.type)  ._undefined }}
```
Comment (choldgraf) [inline jupyterhub/templates/proxy/ingress-internal.yaml]: just a note to change this if `files` changes to `manual` or whatever else

**Comment by choldgraf** [comment]
as a non-letsencrypt person, files and manual are equally non-informative to me, just FYI. I'm 33/33/33 on files/manual/something else.

**Comment by yuvipanda** [comment]
@minrk yeah, I think that'd be too much magic. Plus, I think there's a good pattern of:

```
type: someString
someString:  
   <someString specific options>
```

This occurs throughout the k8s ecosystem, and is a good one to follow for us.

I agree manual seems not entirely clear, but can't think of anything else.

**Comment by yuvipanda** [comment]
FWIW I also consider provisioning your own certificates an 'advanced' use case.

File: jupyterhub/values.yaml [hunk 4]
```diff
@@ -82,6 +82,8 @@ proxy:
     type: letsencrypt
     letsencrypt:
       contactEmail: ''
+    key:
```
Comment (yuvipanda) [inline jupyterhub/values.yaml]: this should be manual.key and manual.cert IMO. I think we should be doing:

```
type: X
X:
   <all-X-related-options>
```

**Comment by choldgraf** [comment]
agree it's advanced - that's why I'm just giving a datapoint but not weighing in heavily one way or another :-)

File: jupyterhub/values.yaml (hunk 4 above)
Comment (minrk) [reply jupyterhub/values.yaml]: done

**Pull request merged**